File: 프로그래머스/1/67256. ［카카오 인턴］ 키패드 누르기/［카카오 인턴］ 키패드 누르기.java
class Point {
    public int x;
    public int y;
    public char hand;
    
    public Point(int number, char hand) {
        this.x = (number + 2) / 3;
        this.y = (number + 2) % 3;
        this.hand = hand;
    }
    
    public char moveTo(int number) {
        x = (number + 2) / 3;
        y = (number + 2) % 3;
        return hand;
    }
    
    public int getDistance(int number) {
        if (number == 0) {
            number = 11;
        }
        
        int x = (number + 2) / 3;
        int y = (number + 2) % 3;
        
        return Math.abs(this.x - x) + Math.abs(this.y - y);
    }
}

class Solution {
    public String solution(int[] numbers, String hand) {
        StringBuilder sb = new StringBuilder();
        Point left = new Point(10, 'L');
        Point right = new Point(12, 'R');
        
        for (int n : numbers) {
            if (n == 0) {
                n = 11;
            }
            if ((n + 2) % 3 == 0) {
                sb.append(left.moveTo(n));
                continue;
            }
            if ((n + 2) % 3 == 2) {
                sb.append(right.moveTo(n));
                continue;
            }
            
            int leftDist = left.getDistance(n);
            int rightDist = right.getDistance(n);
            
            if (leftDist == rightDist) {
                if (hand.equals("left")) {
                    sb.append(left.moveTo(n));
                } else {
                    sb.append(right.moveTo(n));
                }
            } else if (leftDist < rightDist) {
                sb.append(left.moveTo(n));
            } else {
                sb.append(right.moveTo(n));
            }
        }
        
        return sb.toString();
    }
}

File: 프로그래머스/1/67256. ［카카오 인턴］ 키패드 누르기/［카카오 인턴］ 키패드 누르기.py
from math import sqrt


class Point:
    def __init__(self, x, y):
        self.keypad = {
            1: (0, 0), 2: (0, 1), 3: (0, 2),
            4: (1, 0), 5: (1, 1), 6: (1, 2),
            7: (2, 0), 8: (2, 1), 9: (2, 2),
            0: (3, 1),
        }
        self.x = x
        self.y = y
    
    def move_to(self, number):
        nx, ny = self.keypad[number]
        self.x = nx
        self.y = ny
        
    def get_distance(self, number):
        nx, ny = self.keypad[number]
        return abs(self.x - nx) + abs(self.y - ny)
        

def solution(numbers, hand):
    result = []
    left = Point(3, 0)
    right = Point(3, 2)
    
    for n in numbers:
        if n in (1, 4, 7):
            result.append('L')
            left.move_to(n)
            continue
            
        if n in (3, 6, 9):
            result.append('R')
            right.move_to(n)
            continue
            
        left_dist = left.get_distance(n)
        right_dist = right.get_distance(n)

        if left_dist == right_dist:
            if hand == 'right':
                result.append('R')
                right.move_to(n)
            else:
                result.append('L')
                left.move_to(n)
        elif left_dist < right_dist:
            result.append('L')
            left.move_to(n)
        else:
            result.append('R')
            right.move_to(n)
    
    return ''.join(result)